source/_posts/2016-10-24-gsoc-2016-port-nmatrix-to-jruby.markdown

```diff
@@ -0,0 +1,536 @@
+
+```ruby
+def sin
+  result = create_dummy_nmatrix
+  result.s = @s.copy.mapToSelf(Sin.new())
+  result
+end
+```
+
+NMatrix#method(arg) has been implemented using Bivariate functions provided by Commons-Maths API and Java Maths API.
+
+```ruby
+def gamma
+  result = create_dummy_nmatrix
+  result.s = ArrayRealVector.new MathHelper.gamma(@s.toArray)
+  result
+end
+```
+
+```java
+import org.apache.commons.math3.special.Gamma;
+
+public class MathHelper{
+  ...
+  public static double[] gamma(double[] arr){
+    double[] result = new double[arr.length];
+    for(int i = 0; i< arr.length; i++){
+      result[i] = Gamma.gamma(arr[i]);
+    }
+    return result;
+  }
+  ...
+}
+```
+
+## **Decomposition**
+
+NMatrix-MRI relies on LAPACK and ATLAS for matrix decomposition and solving functionalities. Apache Commons Math provides a different set of API for decomposing a matrix and solving an equation. For example, `#potrf` and other LAPACK specific functions have not been implemented as they are not required at all.
+
+Calculating determinant in NMatrix is tricky where a matrix is reduced either to a Lower or Upper matrix and the Diagonal elements of the matrix are multiplied to get the result. Also, the correct sign of the result (whether positive or negative) is taken into account while calculating the determinant. However, NMatrix-JRuby uses Commons Math API to calculate the determinant.
+
+```ruby
+def det_exact
+  if (@dim != 2 || @shape[0] != @shape[1])
+    raise(ShapeError, "matrices must be square to have a determinant defined")
+    return nil
+  end
+  to_return = LUDecomposition.new(self.twoDMat).getDeterminant()
+end
+```
+
+Given below is the code, that shows how Cholesky decomposition has been implemented by using Commons Math API. Similarly, LU Decomposition and QR factorization have been implemented.
+
+**Cholesky Decomposition**
+
+```ruby
+  def factorize_cholesky
+    cholesky = CholeskyDecomposition.new(self.twoDMat)
+    l = create_dummy_nmatrix
+    twoDMat = cholesky.getL
+    l.s = ArrayRealVector.new(ArrayGenerator.getArrayDouble\
+        (twoDMat.getData, @shape[0], @shape[1]))
+
+    u = create_dummy_nmatrix
+    twoDMat = cholesky.getLT
+    u.s = ArrayRealVector.new(ArrayGenerator.getArrayDouble\
+      (twoDMat.getData, @shape[0], @shape[1]))
+    return [u,l]
+  end
+```
+
+Cholesky Decomposition for an NMatrix-JRuby requires the matrix to be square matrix.
+
+**LUDecomposition**
+
+```ruby
+  def factorize_lu with_permutation_matrix=nil
+    raise(NotImplementedError, "only implemented for dense storage")\
+       unless self.stype == :dense
+    raise(NotImplementedError, "matrix is not 2-dimensional")\
+       unless self.dimensions == 2
+    t = self.clone
+    pivot = create_dummy_nmatrix
+    twoDMat = LUDecomposition.new(self.twoDMat).getP
+    pivot.s = ArrayRealVector.new(ArrayGenerator.getArrayDouble\
+    (twoDMat.getData, @shape[0], @shape[1]))
+    return [t,pivot]
+  end
+```
+
+**QRFactorization**
```

Comment:
All of these sections are not needed if they're just showing code. Instead use a bullet list with links to the code, showing something like:
* Cholesky decomposition
* LU decomposition
* QR decomposition